### PR TITLE
Feature/cc 0104 make payment and complete order

### PR DESCRIPTION
**Pull request by STY1997**
.env
#Stripe
STRIPE_SECRET_KEY=
STRIPE_PUBLISHABLE_KEY=
STRIPE_WEBHOOK_SECRET_KEY=
DOMAIN=http://localhost:3000

**Stripe Local Environment set up:**
via Browser
1. Sign up a stripe account
2. At the top right of the dashboard, tick **test mode** and click **developers**
Screen Shot 2022-11-08 at 17 49 42

3. On the left hand side, click **API keys**, then you can find the **secret key** and **publishable key**
4. Below API keys, click **Webhooks,** at bottom right, click **add local listener**

Screen Shot 2022-11-08 at 17 54 46

via Terminal
6. Follow the instruction to set up the local environment, make sure stripe is listen to "localhost:8080/v1/stripe/webhook"
https://dashboard.stripe.com/test/webhooks/create?endpoint_location=local

Screen Shot 2022-11-08 at 17 58 10

7. If every thing goes well, you will get the **webhook secret key**(also called **endpointsSecret**)
8. Fill all three keys in .env and don't forget the DOMAIN

*Also do not forget to run **npm i**

### CODE REVIEW

File: src/stripe/stripe.service.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  ): Promise<{ paymentInfo?: PaymentInfo; order: Order }> {
+    const order = await this.orderService.findOne(id);
+    const paymentInfo = await this.paymentInfoModel.findOne({ orderId: id }).exec();
+    if (!order) throw new NotFoundException(`order #${id} not found`);
+    return {
+      paymentInfo,
+      order,
+    };
+  }
+
```
Comment (RedRe4per) [inline src/stripe/stripe.service.ts:97]: My suggestion is that it's better to add a payment ref in order schema, and delete "findPaymentInfoByOrderId". Instead, you can use order.findById(id).populate().exec() to get an order's payment data.

There are two advantages to doing this:
1. When you search for multiple orders, you can get the payment data corresponding to all orders through .populate(). Without this association, it cannot be done. This feature will be used on the my order page.
2. When you later want to filter orders by attributes such as payment status/construction address, you must have this association to filter through the sub-table payment.

**Review by RedRe4per — CHANGES_REQUESTED** [review]

File: src/orders/dto/updateOrder.dto.ts [hunk 2]
```diff
@@ -1,13 +1,9 @@
...
 import { CreateOrderDto } from "./createOrder.dto";
 
-export class UpdateOrderDto extends PartialType(CreateOrderDto) {
-  @Exclude()
-  user_id: string;
-
+export class UpdateOrderDto extends PartialType(OmitType(CreateOrderDto, ["user_id"] as const)) {
   @IsEnum(StatusType)
   @IsNotEmpty()
   status: string;
```
[the remark this answers is not in the document]
Comment (STY1997) [reply src/orders/dto/updateOrder.dto.ts:9]: > The order state cannot be modified by the frontend. please delete.

fixed

File: src/stripe/dto/create-payment-information.dto.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  @IsNotEmpty()
+  state: string;
+}
+
+export class CreatePaymentInfoDto {
+  @IsString()
+  @IsNotEmpty()
+  readonly orderId: string;
+
+  @IsString()
```
[the remark this answers is not in the document]
Comment (STY1997) [reply src/stripe/dto/create-payment-information.dto.ts:32]: > @IsEmail()

fixed

File: src/stripe/dto/create-payment-information.dto.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  @IsNotEmpty()
+  readonly currency: string;
+
+  @IsString()
+  @IsNotEmpty()
+  readonly sessionId: string;
+
+  @IsString()
+  @IsNotEmpty()
+  readonly amountTotal: string;
```
[the remark this answers is not in the document]
Comment (STY1997) [reply src/stripe/dto/create-payment-information.dto.ts:54]: > @IsNumber()

fixed

File: src/stripe/stripe.controller.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+          },
+        },
+      };
+    });
+
+    try {
+      const session = await this.stripeClient.checkout.sessions.create({
+        line_items,
+        metadata: { orderId: order_Id },
+        payment_method_types: ["card"],
```
[the remark this answers is not in the document]
Comment (STY1997) [reply src/stripe/stripe.controller.ts:54]: > * "afterpay"

fixed

File: src/stripe/stripe.controller.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  @Get("paymentInfo/:orderId")
+  async getPaymentInfoAndOrderByOrderId(
+    @Param("orderId") orderId: ObjectId,
+  ): Promise<{ paymentInfo?: PaymentInfo; order: Order }> {
+    return await this.stripeService.findPaymentInfoByOrderId(orderId);
+  }
+
+  @Post("paymentInfo/:orderId")
```
[the remark this answers is not in the document]
Comment (STY1997) [reply src/stripe/stripe.controller.ts:75]: > @put() or @patch()

fixed

File: src/stripe/stripe.service.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  ) {}
+
+  /**
+   * handle stripe checkout complete event.
+   * @param event stripe webhook event
+   */
+  @StripeWebhookHandler("checkout.session.completed")
+  stripeWebhookHandler(event: Stripe.Event) {
+    const dataObject = event.data.object as Stripe.Checkout.Session;
+    console.log(`order:${dataObject.metadata.orderId} payment completed`);
```
[the remark this answers is not in the document]
Comment (STY1997) [reply src/stripe/stripe.service.ts:27]: > console.log remove

fixed

**Review by RedRe4per — APPROVED** [review]

File: src/stripe/stripe.service.ts (hunk 1 above)
Comment (STY1997) [reply src/stripe/stripe.service.ts:103]: > My suggestion is that it's better to add a payment ref in order schema, and delete "findPaymentInfoByOrderId". Instead, you can use order.findById(id).populate().exec() to get an order's payment data.
> 
> There are two advantages to doing this:
> 
> 1. When you search for multiple orders, you can get the payment data corresponding to all orders through .populate(). Without this association, it cannot be done. This feature will be used on the my order page.
> 2. When you later want to filter orders by attributes such as payment status/construction address, you must have this association to filter through the sub-table payment.

fixed, check the findAll function in the order provider (orders/order.service.ts)

**Review by ShawnWang86b — APPROVED** [review]

**Pull request merged**